p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,790 @@
+		}
+	}
+
+	return &streamRoundTripper{h: h.StreamHost, server: server.ID, httpHost: h}, nil
+}
+
+type httpMultiaddr struct {
+	useHTTPS bool
+	host     string
+	port     string
+	sni      string
+}
+
+func parseMultiaddr(addr ma.Multiaddr) httpMultiaddr {
+	out := httpMultiaddr{}
+	ma.ForEach(addr, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case ma.P_IP4, ma.P_IP6, ma.P_DNS, ma.P_DNS4, ma.P_DNS6:
+			out.host = c.Value()
+		case ma.P_TCP, ma.P_UDP:
+			out.port = c.Value()
+		case ma.P_TLS, ma.P_HTTPS:
+			out.useHTTPS = true
+		case ma.P_SNI:
+			out.sni = c.Value()
+
+		}
+		return out.host == "" || out.port == "" || !out.useHTTPS || out.sni == ""
+	})
+
+	if out.useHTTPS && out.sni == "" {
+		out.sni = out.host
+	}
+	return out
+}
+
+var httpComponent, _ = ma.NewComponent("http", "")
+var tlsComponent, _ = ma.NewComponent("tls", "")
+
+// normalizeHTTPMultiaddr converts an https multiaddr to a tls/http one.
+// Returns a bool indicating if the input multiaddr has an http (or https) component.
+func normalizeHTTPMultiaddr(addr ma.Multiaddr) (ma.Multiaddr, bool) {
+	isHTTPMultiaddr := false
+	beforeHTTPS, afterIncludingHTTPS := ma.SplitFunc(addr, func(c ma.Component) bool {
+		if c.Protocol().Code == ma.P_HTTP {
+			isHTTPMultiaddr = true
+		}
+
+		if c.Protocol().Code == ma.P_HTTPS {
+			isHTTPMultiaddr = true
+			return true
+		}
+		return false
+	})
+
+	if afterIncludingHTTPS == nil {
+		// No HTTPS component, just return the original
+		return addr, isHTTPMultiaddr
+	}
+
+	_, afterHTTPS := ma.SplitFirst(afterIncludingHTTPS)
+	if afterHTTPS == nil {
+		return ma.Join(beforeHTTPS, tlsComponent, httpComponent), isHTTPMultiaddr
+	}
+
+	return ma.Join(beforeHTTPS, tlsComponent, httpComponent, afterHTTPS), isHTTPMultiaddr
+}
+
+// ProtocolPathPrefix looks up the protocol path in the well-known mapping and
+// returns it. Will only store the peer's protocol mapping if the server ID is
+// provided.
+func (h *Host) getAndStorePeerMetadata(roundtripper http.RoundTripper, server peer.ID) (PeerMeta, error) {
+	if h.peerMetadata == nil {
+		h.peerMetadata = newPeerMetadataCache()
+	}
+	if meta, ok := h.peerMetadata.Get(server); server != "" && ok {
+		return meta, nil
+	}
+
+	req, err := http.NewRequest("GET", "/.well-known/libp2p", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+
+	client := http.Client{Transport: roundtripper}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body := [peerMetadataLimit]byte{}
+	bytesRead := 0
+	for {
+		n, err := resp.Body.Read(body[bytesRead:])
+		bytesRead += n
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if bytesRead >= peerMetadataLimit {
+			return nil, fmt.Errorf("peer metadata too large")
+		}
+	}
```

Comment:
Consider using an `io.LimitedReader`. Avoids allocating 8 kB, which won't be needed in the vast majority of cases.